tools/rewriter/SourceRewriter.cpp

```diff
@@ -77,6 +77,7 @@ struct DirectoryParser : public llvm::cl::parser<std::string> {
       llvm::errs() << ec.message() << '\n';
       return true;
     }
+    llvm::sys::path::remove_dots(dir, true);
```

Comment:
Lol LLVM just calls this function `remove_dots`?

Reply:
Does this do canonicalization or just remove the dots? Canonicalization also needs to resolve links.

Reply:
No this doesn't fully canonicalize, it just removes `.` and `..` from the path. I couldn't find a proper canonicalize function, at least in LLVM. If we really want full canonicalization we may need to use the stdlib one (unless someone else can find an LLVM function I'm missing).

Reply:
Let's just use the `std::filesystem` one then.

Reply:
Might be better to copy the LLVM approach of `make_absolute` followed by `remove_dots`. They should give the same thing but we're comparing it against paths generated by LLVM so that seems slightly less likely to create weird bugs.

Reply:
`make_absolute` and `remove_dots` won't resolve symlinks, though.  If we compare it to an LLVM path, we should also fully resolve and canonicalize it.

I think I found the LLVM canonicalization function: [`llvm::sys::fs::real_path`](https://llvm.org/doxygen/namespacellvm_1_1sys_1_1fs.html#a2363fbfbd0a348f5d81fc3d3223ecae3).

Reply:
Oh dope, it does exist it just doesn't at all have the name I expected. In that case I think using `real_path` is the right thing to do here.